modules/3d/test/test_normal.cpp

```diff
@@ -396,12 +369,305 @@ class RgbdPlaneTest
+        split(dstNormals, dstNormalsCh);
+        dstNormalsCh.resize(3);
+        merge(dstNormalsCh, newDstNormals);
+        dstNormals = newDstNormals;
+
+        Mat dstMask(dstNormals.size(), CV_8U);
+        for (int y = 0; y < dstMask.rows; y++)
+        {
+            uchar* dstMaskRow = dstMask.ptr<uchar>(y);
+            if (depth == CV_32F)
+            {
+                Vec3f *dstrow = dstNormals.ptr<Vec3f>(y);
+                for (int x = 0; x < dstMask.cols; x++)
+                {
+                    dstMaskRow[x] = (dstrow[x] == dstrow[x])*255;
+                }
+            }
+            else if (depth == CV_64F)
+            {
+                Vec3d *dstrow = dstNormals.ptr<Vec3d>(y);
+                for (int x = 0; x < dstMask.cols; x++)
+                {
+                    dstMaskRow[x] = (dstrow[x] == dstrow[x])*255;
+                }
+            }
+        }
+        // div by 8 because uchar is 8-bit
+        res.maskl2 = cv::norm(dstMask, srcMask, NORM_HAMMING)/8;
+
+        // Flipping Y and Z to correspond to srcNormals
+
+        for (int y = 0; y < dstNormals.rows; y++)
+        {
+            if (depth == CV_32F)
+            {
+                Vec3f *nrmrow = dstNormals.ptr<Vec3f>(y);
+                for (int x = 0; x < dstNormals.cols; x++)
+                {
+                    Vec3f n = nrmrow[x];
+                    n[1] = -n[1];
+                    n[2] = -n[2];
+                    nrmrow[x] = n;
+                }
+            }
+            else if (depth == CV_64F)
+            {
+                Vec3d *nrmrow = dstNormals.ptr<Vec3d>(y);
+                for (int x = 0; x < dstNormals.cols; x++)
+                {
+                    Vec3d n = nrmrow[x];
+                    n[1] = -n[1];
+                    n[2] = -n[2];
+                    nrmrow[x] = n;
+                }
+            }
+        }
+
+        Mat absdot(1080, 1920, depth, Scalar(0));
+        for (int y = 0; y < srcNormals.rows; y++)
+        {
+            if (depth == CV_32F)
+            {
+                Vec3f *srcrow = srcNormals.ptr<Vec3f>(y);
+                Vec3f *dstrow = dstNormals.ptr<Vec3f>(y);
+                float *absrow = absdot.ptr<float>(y);
+                for (int x = 0; x < srcNormals.cols; x++)
+                {
+                    Vec3f sn = srcrow[x];
+                    Vec3f dn = dstrow[x];
+                    absrow[x] = std::abs(sn.dot(dn));
+                }
+            }
+            else if (depth == CV_64F)
+            {
+                Vec3d *srcrow = srcNormals.ptr<Vec3d>(y);
+                Vec3d *dstrow = dstNormals.ptr<Vec3d>(y);
+                double *absrow = absdot.ptr<double>(y);
+                for (int x = 0; x < srcNormals.cols; x++)
+                {
+                    Vec3d sn = srcrow[x];
+                    Vec3d dn = dstrow[x];
+                    absrow[x] = std::abs(sn.ddot(dn));
+                }
+            }
+        }
```

Comment:
Looks like `cv::Mat::dot` Does what you what, if each matrix pixel has 3 channels of float of double type. You need just apply abs to the result matrix. Also it makes sense to add comment how the dot product is calculated to make the idea more obvious.

Reply:
But what I see is that `cv::Mat::dot` returns one number instead of array. Also it does not support masks.